mono/metadata/icall-table.h

```diff
@@ -24,4 +24,165 @@ mono_install_icall_table_callbacks (MonoIcallTableCallbacks *cb);
 MONO_API void
 mono_icall_table_init (void);
 
+// All icall parameter and return types need a handler macro here.
+#define MONO_IF_NOT_VOID(type, x) MONO_IF_NOT_VOID_##type (x)
+#define MONO_IF_NOT_VOID_void(x) /* nothing */
+#define MONO_IF_NOT_VOID_int(x) x
+#define MONO_IF_NOT_VOID_uint(x) x
+#define MONO_IF_NOT_VOID_guint32(x) x
+#define MONO_IF_NOT_VOID_gint32(x) x
+#define MONO_IF_NOT_VOID_gint64(x) x
+#define MONO_IF_NOT_VOID_MonoObject(x) x
+#define MONO_IF_NOT_VOID_MonoString(x) x
+#define MONO_IF_NOT_VOID_gpointer(x) x
+#define MONO_IF_NOT_VOID_MonoBoolean(x) x
+#define MONO_IF_NOT_VOID_MonoW32ProcessInfo(x) x
+
+// All icall parameter and return types need a handler macro here.
+#define MONO_IF_VOID(type, x) MONO_IF_VOID_##type (x)
+#define MONO_IF_VOID_void(x) x
+#define MONO_IF_VOID_int(x) /* nothing */
+#define MONO_IF_VOID_guint32(x) /* nothing */
+#define MONO_IF_VOID_gint32(x) /* nothing */
+#define MONO_IF_VOID_gint64(x) /* nothing */
+#define MONO_IF_VOID_MonoObject(x) /* nothing */
+#define MONO_IF_VOID_MonoString(x) /* nothing */
+#define MONO_IF_VOID_gpointer(x) /* nothing */
+#define MONO_IF_VOID_MonoBoolean(x) /* nothing */
+#define MONO_IF_VOID_MonoW32ProcessInfo(x) /* nothing */
+
+// All icall parameter and return types need a handler macro here.
+#define MONO_IF_OBJECT(type, x) MONO_IF_OBJECT_##type (x)
+#define MONO_IF_OBJECT_void(x) /* nothing */
+#define MONO_IF_OBJECT_int(x) /* nothing */
+#define MONO_IF_OBJECT_guint32(x) /* nothing */
+#define MONO_IF_OBJECT_gint32(x) /* nothing */
+#define MONO_IF_OBJECT_gint64(x) /* nothing */
+#define MONO_IF_OBJECT_gpointer(x) /* nothing */
+#define MONO_IF_OBJECT_MonoBoolean(x) /* nothing */
+#define MONO_IF_OBJECT_MonoObject(x) x
+#define MONO_IF_OBJECT_MonoString(x) x
+#define MONO_IF_OBJECT_MonoW32ProcessStartInfo(x) x
+#define MONO_IF_OBJECT_MonoW32ProcessInfo(x) /* TODO */
+
+// All icall parameter and return types need a handler macro here.
+#define MONO_IF_NOT_OBJECT(type, x) MONO_IF_NOT_OBJECT_##type (x)
```

Comment:
If this is going to go forward, I think this `MONO_IF_XXX` idea won't scale well - need some case/switch macros, not just booleans.

Because there are actually 5 different ways to pass arguments to icalls - just "object/not object" is not enough:
https://github.com/mono/mono/blob/d8a5d891be97c333fa13b6ca2918a5c0a829c729/mono/metadata/marshal-ilgen.c#L6160-L6176

1. No wrapping - roughly `MONO_IF_NOT_OBJECT` - this is for primitives.
2. Wrap object - roughly `MONO_IF_OBJECT` - this is pointers to managed objects.
3. Wrap object inout - for reference types passed by `ref` - we copy back the value from the handle when the icall returns
4. Wrap object out - for reference types passed by `out` - we initialize the handle with null (this is needed because the `out` param usually has garbage in it, but GC scanning of coop handles is precise, so garbage leads to crashes).
5. Valuetype ref - for value types passed by `out` or `ref` - we pin the address of the valuetype using an interior coop handle (this is pretty much the only place where we use them) and then pass the _raw pointer_ to managed.  This is for things like `MonoW32ProcessStartInfo` and the various reflection `Info` structs, and so on.